File: app/models/content_submission.py
"""Pydantic models for content submission workflows."""

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, HttpUrl

from app.models.metadata import ContentStatus, ContentType


class SubmitContentRequest(BaseModel):
    """Request to submit a user-provided URL for processing."""

    url: HttpUrl = Field(..., description="URL to submit (http/https only)")
    content_type: ContentType | None = Field(
        None,
        description="Content type hint. If omitted, the server will infer based on the URL.",
    )
    title: str | None = Field(
        None,
        max_length=500,
        description="Optional title supplied by the client/share sheet",
    )
    platform: str | None = Field(
        None, max_length=50, description="Optional platform hint (e.g., spotify, substack)"
    )
    instruction: str | None = Field(
        None,
        max_length=4000,
        validation_alias=AliasChoices("instruction", "note"),
        description="Optional instruction for analyzing the submitted URL",
    )
    crawl_links: bool = Field(
        False,
        description=(
            "Whether to create additional content items from relevant links "
            "discovered on the submitted page."
        ),
    )
    subscribe_to_feed: bool = Field(
        False,
        description=(
            "When true, detect an RSS/Atom feed from the URL and subscribe to it "
            "instead of processing the URL as content."
        ),
    )
    share_and_chat: bool = Field(
        False,
        description=(
            "When true, mark the submitted content as read and start a dig-deeper chat "
            "after processing completes."
        ),
    )
    save_to_knowledge_and_mark_read: bool = Field(
        False,
        description=(
            "When true, download and summarize the submitted content, then mark it as "
            "read and save it to the user's knowledge library."
        ),
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "url": "https://open.spotify.com/episode/abc123",
                "content_type": "podcast",
                "title": "Great interview about AI",
                "platform": "spotify",
                "instruction": "Add all links mentioned in the episode page",
                "crawl_links": True,
                "subscribe_to_feed": False,
                "share_and_chat": False,
                "save_to_knowledge_and_mark_read": False,
            }
        }
    )


class ContentSubmissionResponse(BaseModel):
    """Response describing the result of a user submission."""

    content_id: int = Field(..., description="ID of the created or existing content")
    content_type: ContentType = Field(..., description="Content type that will be processed")
    status: ContentStatus = Field(..., description="Current processing status of the content")
    platform: str | None = Field(None, description="Normalized platform name if available")
    already_exists: bool = Field(
        False, description="Whether the submission matched an existing record"
    )
    message: str = Field(..., description="Human-readable status message")
    task_id: int | None = Field(None, description="Processing task ID enqueued for this content")
    source: str | None = Field(
        None, description="Source attribution recorded for the content (self submission)"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "content_id": 42,
                "content_type": "podcast",
                "status": "new",
                "platform": "spotify",
                "already_exists": False,
                "message": "Content queued for processing",
                "task_id": 101,
                "source": "self submission",
            }
        }
    )

File: rust/crates/newsly-contracts/src/content_submission.rs
use std::fmt::{self, Display, Formatter};
use std::str::FromStr;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

macro_rules! string_enum {
    ($name:ident { $($variant:ident => $value:literal),+ $(,)? }) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
        #[serde(rename_all = "snake_case")]
        pub enum $name {
            $($variant),+
        }

        impl $name {
            pub const fn as_str(self) -> &'static str {
                match self {
                    $(Self::$variant => $value),+
                }
            }
        }

        impl Display for $name {
            fn fmt(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
                formatter.write_str(self.as_str())
            }
        }

        impl FromStr for $name {
            type Err = UnknownContentValue;

            fn from_str(value: &str) -> Result<Self, Self::Err> {
                match value {
                    $($value => Ok(Self::$variant)),+,
                    _ => Err(UnknownContentValue {
                        kind: stringify!($name),
                        value: value.to_owned(),
                    }),
                }
            }
        }
    };
}

string_enum!(ContentType {
    Article => "article",
    Podcast => "podcast",
    News => "news",
    InsightReport => "insight_report",
    Unknown => "unknown",
});

string_enum!(ContentStatus {
    New => "new",
    Pending => "pending",
    Processing => "processing",
    AwaitingImage => "awaiting_image",
    Completed => "completed",
    Failed => "failed",
    Skipped => "skipped",
});

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
/// Request to submit a user-provided URL for processing.
#[allow(clippy::struct_excessive_bools)] // The public wire contract exposes independent actions.
pub struct SubmitContentRequest {
    /// URL to submit (HTTP/HTTPS only).
    pub url: String,
    /// Optional content-type hint. New records are still persisted as unknown until analysis.
    pub content_type: Option<ContentType>,
    /// Optional title supplied by the client or share sheet.
    #[schemars(length(max = 500))]
    #[schema(max_length = 500)]
    pub title: Option<String>,
    /// Optional platform hint, such as `spotify` or `substack`.
    #[schemars(length(max = 50))]
    #[schema(max_length = 50)]
    pub platform: Option<String>,
    /// Optional analysis instruction. `note` remains accepted as the legacy input alias.
    #[serde(alias = "note")]
    #[schemars(length(max = 4_000))]
    #[schema(max_length = 4_000)]
    pub instruction: Option<String>,
    /// Whether analysis should create content items for relevant discovered links.
    #[serde(default)]
    pub crawl_links: bool,
    /// Whether to detect and subscribe to a feed instead of saving the URL as inbox content.
    #[serde(default)]
    pub subscribe_to_feed: bool,
    /// Whether to mark the content read and start a dig-deeper chat when it is ready.
    #[serde(default)]
    pub share_and_chat: bool,
    /// Optional first user message for the share-and-chat session.
    #[schemars(length(max = 2_000))]
    #[schema(max_length = 2_000)]
    pub chat_initial_message: Option<String>,
    /// Whether to mark the content read and save it to the user's Knowledge library.
    #[serde(default)]
    pub save_to_knowledge_and_mark_read: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
/// Result of creating or reusing one submitted URL.
pub struct ContentSubmissionResponse {
    /// Created or reused Content identifier.
    pub content_id: i64,
    /// Current durable Content type.
    pub content_type: ContentType,
    /// Current durable processing status.
    pub status: ContentStatus,
    /// Normalized platform for an existing record; new submissions return null until analysis.
    pub platform: Option<String>,
    /// Whether the URL matched an existing Content row.
    #[serde(default)]
    pub already_exists: bool,
    /// Human-readable submission result.
    pub message: String,
    /// Reused or newly queued analysis task identifier, when analysis is needed.
    pub task_id: Option<i64>,
    /// Durable source attribution.
    pub source: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UnknownContentValue {
    kind: &'static str,
    value: String,
}

impl Display for UnknownContentValue {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
        write!(formatter, "unknown {} value {:?}", self.kind, self.value)
    }
}

impl std::error::Error for UnknownContentValue {}

#[cfg(test)]
mod tests {
    use super::SubmitContentRequest;

    #[test]
    fn note_is_an_input_alias_for_instruction() {
        let request: SubmitContentRequest = serde_json::from_value(serde_json::json!({
            "url": "https://example.com/article",
            "note": "Follow every link"
        }))
        .unwrap();

        assert_eq!(request.instruction.as_deref(), Some("Follow every link"));
        assert!(!request.crawl_links);
        assert!(!request.subscribe_to_feed);
    }
}
